Adjustments to nimble run

atstp opened the pull request:
this should address #767

atstp commented:
I also see that you guys have tests for some issues in `tests/issueNUMBER`.

If you guys feel like these changes should be tested, i can add to this pull request.

dom96 commented:
Seems good. although CI is failing, can you fix?

dom96 commented:
Also, please test with some escaped characters in your shell, Nim's `escape` might do some weird things for it.

atstp commented:
I'll fix the failing CI and add tests. I opened this PR early to get feedback on the changes before I shore up stuff with tests.

I'll also add tests for escaped characters, per your suggestion.

Are you okay with `nimble run` re-using the exit code of the binary?

alaviss commented on src/nimble.nim:
```diff
@@ -1129,10 +1129,8 @@ proc run(options: Options) =
-  let args = options.action.runFlags.join(" ")
-
-  doCmd("$# $#" % [binaryPath, args], showOutput = true)
+  let args = options.action.runFlags.map(x => x.escape).join(" ")
```
```suggestion
  let args = options.action.runFlags.map(x => x.quoteShell).join(" ")
```

`os.quoteShell` is much better for quoting things to pass to the shell.

atstp replied:
ah, thanks. I hadn't seen that proc.

Looking at the docs, I also now see `quoteShellCommand`. I'm assuming it would be appropriate to use that instead of the current `quoteShell`, `join`, and `%`?

atstp commented:
I'm looking at the details of the most recent CI failure on this PR. It looks like that's job [`806`](https://travis-ci.org/nim-lang/nimble/builds/648757270), and the failures that i saw revolved around `travis.sh` expecting input:

```
$ source travis.sh
    Prompt: Symlink for 'nimble' detected in '/home/travis/.nimble/bin'. Can I remove it? [y/N]
    Answer: 

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

The build has been terminated
```

Could someone more familiar with nimble's CI take a look? If this is my mistake I'd like to get things fixed.

genotrance commented:
@atstp - fix has been merged, can you please rebase this PR and retry?

atstp commented:
@genotrance, thanks, that fixed it

dom96 commented on tests/tester.nim:
```diff
@@ -925,56 +925,58 @@ test "Passing command line arguments to a task (#633)":
       check exitCode == QuitSuccess
-      check output.contains("tests$1run$1$2 --debug check" %
-                            [$DirSep, "run".changeFileExt(ExeExt)])
-      check output.contains("""Testing `nimble run`: @["--debug", "check"]""")
```
I guess you wanted to consolidate everything into showCliParams, but I'd rather we would have both of these tests. That way we ensure your changes don't break anything. Please restore these. It's fine to have `run` and `showCliParams` packages.

atstp replied:
Okay, all the changes related to the `run` package are removed

atstp commented:
Also, on `master`, three of the four tests in the "nimble run" suite contained the `--debug` flag and followed by a check that the debugging output included the command that was run:

```
     var (output, exitCode) = execNimble(
        "--debug",
        "run"
# ...
      )
# ... 
      check output.contains("tests$1run$1$2 ..." %
                            [$DirSep, "run".changeFileExt(ExeExt)])
```

To simplify things, I removed the triplicates and added a test to check for that specifically.

If you guys feel like that's too much of a change to the test suite, let me know and I'll modify the PR.

dom96 commented:
I'd rather you keep the tests as they are. If you want to simplify them then do so in another PR, easier to review that way.

atstp commented:
sounds good, it should be good to go now
